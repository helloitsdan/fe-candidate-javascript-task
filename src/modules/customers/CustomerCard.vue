<template>
  <div class="card">
    <h3 class="is-marginless">{{ name | formatName }}</h3>

    <p>{{ email }}</p>

    <p>
      <strong>Registered on</strong>
      {{ registered | moment("dddd, MMMM Do YYYY") }}
    </p>

    <p v-if="!isActive">
      This customer is
      <strong>inactive.</strong>
    </p>

    <p class="is-marginless">
      <router-link
        :to="{ name: 'CustomerDetails', params: { customerId: uuid } }"
      >
        View details
      </router-link>
    </p>
  </div>
</template>

<script lang="ts">
import Vue from "vue";

export default Vue.extend({
  name: "CustomerCard",
  props: {
    uuid: String,
    isActive: Boolean,
    name: Object,
    email: String,
    registered: String
  }
});
</script>
