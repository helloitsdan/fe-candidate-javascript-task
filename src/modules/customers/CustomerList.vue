<template>
  <div class="background--barely">
    <section class="container padded-block">
      <h1 class="title-margin">Customers</h1>

      <div v-if="loading" class="loading">
        Loading...
      </div>

      <div v-if="error" class="error">
        Sorry! We couldn't load the list of customers right now.
      </div>

      <div v-if="customers" class="columns is-multiline">
        <div
          v-for="customer in customers"
          v-bind:key="customer.uuid"
          class="column is-half"
        >
          <customer-card class="background--white" v-bind="customer" />
        </div>
      </div>
    </section>
  </div>
</template>

<script lang="ts">
import Vue from "vue";

import getCustomers from "@/services/customers";
import CustomerCard from "./CustomerCard.vue";

export default Vue.extend({
  name: "CustomerList",
  components: {
    CustomerCard
  },
  data() {
    return {
      loading: false,
      error: null,
      customers: []
    };
  },
  created() {
    this.fetchCustomerList();
  },
  methods: {
    async fetchCustomerList() {
      if (this.loading) {
        return;
      }

      this.loading = true;

      try {
        this.customers = await getCustomers();
      } catch (e) {
        this.error = e;
      }

      this.loading = false;
    }
  }
});
</script>
