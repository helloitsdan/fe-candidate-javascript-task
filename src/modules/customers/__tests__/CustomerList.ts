import Vue from "vue";
import { shallowMount } from "@vue/test-utils";

import CustomerList from "../CustomerList.vue";
import CustomerCard from "../CustomerCard.vue";
import customerRequest from "@/services/customers";

jest.mock("@/services/customers");

describe("CustomerList", () => {
  const customerRequestMock = customerRequest as jest.Mock;

  const customerList = [
    {
      uuid: "1234",
      isActive: true,
      name: {
        first: "Test",
        last: "Customer"
      },
      email: "test.user@pretend.biz",
      registered: "2016-07-13T02:57:55.235Z"
    },
    {
      uuid: "5678",
      isActive: false,
      name: {
        first: "Second",
        last: "Customer"
      },
      email: "second.user@pretend.biz",
      registered: "2016-09-13T02:57:55.235Z"
    }
  ];

  it("renders", async () => {
    customerRequestMock.mockReturnValue(Promise.resolve(customerList));
    const wrapper = shallowMount(CustomerList);

    await Vue.nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders a CustomerCard for every customer", async () => {
    customerRequestMock.mockReturnValue(Promise.resolve(customerList));
    const wrapper = shallowMount(CustomerList);

    await Vue.nextTick();

    expect(wrapper.findAll(CustomerCard).length).toEqual(customerList.length);
  });

  it("renders a loading message when the request is still pending", () => {
    const wrapper = shallowMount(CustomerList);
    expect(wrapper.find(".loading").exists()).toBe(true);
  });

  it("renders an error message when the request fails", async () => {
    customerRequestMock.mockReturnValue(Promise.reject(":("));
    const wrapper = shallowMount(CustomerList);

    await Vue.nextTick();

    expect(wrapper.find(".error").exists()).toBe(true);
  });
});
