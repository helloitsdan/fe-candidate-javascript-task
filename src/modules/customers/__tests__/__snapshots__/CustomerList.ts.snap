// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerList renders 1`] = `
<div
  class="background--barely"
>
  <section
    class="container padded-block"
  >
    <h1
      class="title-margin"
    >
      Customers
    </h1>
     
    <!---->
     
    <!---->
     
    <div
      class="columns is-multiline"
    >
      <div
        class="column is-half"
      >
        <customer-card-stub
          class="background--white"
          email="test.user@pretend.biz"
          isactive="true"
          name="[object Object]"
          registered="2016-07-13T02:57:55.235Z"
          uuid="1234"
        />
      </div>
      <div
        class="column is-half"
      >
        <customer-card-stub
          class="background--white"
          email="second.user@pretend.biz"
          name="[object Object]"
          registered="2016-09-13T02:57:55.235Z"
          uuid="5678"
        />
      </div>
    </div>
  </section>
</div>
`;
